lib/utils/fs.go

```diff
@@ -233,11 +233,11 @@ func overwriteFile(filePath string) error {
 	if err != nil {
 		return trace.ConvertSystemError(err)
 	}
-	defer f.Close()
 
 	fi, err := f.Stat()
 	if err != nil {
-		return trace.Wrap(err)
+		f.Close()
```

Comment:
Check the Close error too, that's the point of closing explicitly.

I would do this in the same line the defer was, outside of this block.

Reply:
Are you thinking to just log the Close error here or wrap the original error if it isn't nil?

Reply:
Sorry, I misread the snippet initially. Either logging at a low level or swallowing is fine by me here (although, as a team practice, we tend not to swallow).

A helper method or something like the pattern below could do the job:

```go
// After the os.OpenFile block:
closeSilently := true
defer func() {
	if !closeSilently {
		return
	}
	if err := f.Close(); err != nil {
		// TODO Log error here
	}
}()

// Rest of the body goes here.
// (...)

// At the end:
closeSilently = false
return trace.ConvertSystemError(f.Close())
```

Reply:
Got it. I went ahead and just added a log as that seems to be consistent with other parts of the code I found.

Reply:
Let's keep the explicit f.Close() we had at the end. Because we are writing to the file, a Close error could mean we failed to flush, and thus failed to override.

That's why that snippet I sent did all that juggling with `closeSilently`.

Reply:
Makes sense but I'm not loving the housekeeping variable. What do you think about the pattern below? We can use defer so we don't need to track closing the file but we can also return the error if there is no other error and if there is we can log it.

```go
func overwriteFile(filePath string) (err error) {
	defer func() {
		if closeErr := f.Close(); closeErr != nil {
			if err == nil {
				err = trace.ConvertSystemError(closeErr)
			} else {
				log.WithError(closeErr).Warningf("Failed to close %v.", f.Name())
			}
		}
	}()
}
```

Reply:
Sure, that works too.

Reply:
Ok, I pushed up that change.